qiime2/metadata/tests/test_metadata.py

```diff
@@ -312,6 +333,98 @@ def test_missing_data(self):
                                              ('col3', 'categorical'),
                                              ('col4', 'categorical')])
 
+    def test_missing_data_insdc(self):
+        index = pd.Index(['None', 'nan', 'NA', 'foo'], name='id')
+        df = pd.DataFrame(collections.OrderedDict([
+            ('col1', [1.0, np.nan, 'missing', np.nan]),
+            # TODO: it is not currently possible to have an ENTIRELY numeric
+            # column from missing terms, as the dtype of the series is object
+            # and there is not way to indicate the dtype beyond that.
+            # ('NA', [np.nan, np.nan, 'not applicable', np.nan]),
+            ('col3', ['null', 'N/A', 'not collected', 'NA']),
+            ('col4', np.array([np.nan, np.nan, 'restricted access', np.nan],
+                              dtype=object))]),
+            index=index)
+        md = Metadata(df, default_missing_scheme='INSDC:missing')
+
+        self.assertEqual(md.id_count, 4)
+        self.assertEqual(md.column_count, 3)
+        self.assertEqual(md.id_header, 'id')
+        self.assertEqual(md.ids, ('None', 'nan', 'NA', 'foo'))
+        self.assertEqualColumns(md.columns, [('col1', 'numeric'),
+                                             ('col3', 'categorical'),
+                                             ('col4', 'categorical')])
+
+        pd.testing.assert_frame_equal(md.to_dataframe(), pd.DataFrame(
+            {'col1': [1.0, np.nan, np.nan, np.nan],
+             'col3': ['null', 'N/A', np.nan, 'NA'],
+             'col4': np.array([np.nan, np.nan, np.nan, np.nan], dtype=object)},
+            index=index))
+
+    def test_missing_data_insdc_column_missing(self):
+        index = pd.Index(['None', 'nan', 'NA', 'foo'], name='id')
+        df = pd.DataFrame(collections.OrderedDict([
+            ('col1', [1.0, np.nan, 'missing', np.nan]),
+            # TODO: it is not currently possible to have an ENTIRELY numeric
+            # column from missing terms, as the dtype of the series is object
+            # and there is not way to indicate the dtype beyond that.
+            # ('NA', [np.nan, np.nan, 'not applicable', np.nan]),
```

Comment:
The issue is mostly because we cannot provide the type of "categorical" or "numeric" to the constructors for the objects. The current `__init__`s expect the pandas dtype to be sufficient.

This means that a completely NaN column can't be both Object for categorical, and Float for numeric. Since there's a few places where all-nans were expected to be interpreted as categorical and most of the existing tests took that perspective.

It is possible to collapse the object dtype into something more suitable (while preserving object when it's mixed), but that broke existing tests, because all-NaN was always reduced to float. So you'll see a few places where there's a guard to keep the dtype as object.

I think the real solution is to just let `column_types` be a parameter on `__init__`, but I would need to refactor a lot of code to make that work. But if it existed, then it's possible to know that a given series is actually numeric, or categorical, even when it's entirely NaN.

Reply:
I imagine the future structure would look more like:

io.py: reads the TSV as a generic container and provides strings + directives as an intermediate representation
metadata.py: can tolerate literal dtypes or strings, and will parse the strings according to our usual rules

Then the missing value logic can always precede the interpretation, and if the result is entirely NaN, it is still possible to select the correct column type when it is provided. And if not provided, we use the default assumption of Categorical (which is the case right now).

Reply:
Cool, I think we're on the same page here.

> I think the real solution is to just let `column_types` be a parameter on `__init__`

Agreed.